### PR TITLE
Calculate resistances/friction by reducing parallel/series flow segments 

### PR DESCRIPTION
**Pull request by elcojacobs**
This PR overhauls the flow calculation algorithm for the process view.
Flow is now correctly calculated in all cases I think, but more tests should be written.

The new algorithm calculates flow independently for each pressure source (external inputs or pumps) and superimposes them.
It calculates the resistance for the total path by reducing series and parallel segments. With that total resistance, it can calculate the start flow. On splits, this flow is divided over the split based on each segment resistance.

There are no source of sink parts after the refactor, just parts that add pressure. Input and outputs go to special coordinate -1,-1. Whether they are input or output is only determined by the pressure and resulting flow. 
We probably merge them into a single part.

### CODE REVIEW

File: src/plugins/spark/features/ProcessView/calculateFlows.ts [hunk 1]
```diff
@@ -22,6 +25,56 @@ const adjacentPart = (
       !(currentPart && isSamePart(part, currentPart))
       && has(part, ['transitions', outCoords]));
 
+const hasIndependentTransitions = (part: FlowPart): boolean => {
+  // const transitions = partSettings(part).transitions(part);
+  // TODO: calculate this from actual transitions
+  if (part.type === 'BridgeTube' || part.type === 'Pump') {
+    return true;
+  }
+  return false;
```
Comment (steersbob) [inline src/plugins/spark/features/ProcessView/calculateFlows.ts]: - commented code
- the logic in this function literally is `if(boolean value) { return true; } return false;`

File: src/plugins/spark/features/ProcessView/calculateFlows.ts [hunk 2]
```diff
@@ -47,237 +100,237 @@ const additionalFlow = (
   allParts
     .map((item) => {
       if (isSamePart(part, item)) {
-        return { ...item, liquid, calculated: combineFlows(item.calculated, flowToAdd) };
+        return { ...item, liquid, flows: combineFlows(item.flows, flowToAdd) };
       }
       return item;
     });
 
-/*
```
Comment (steersbob) [inline src/plugins/spark/features/ProcessView/calculateFlows.ts:55]: You've removed a lot of documentation, without adding any for the new algorithms.

File: src/plugins/spark/features/ProcessView/calculateFlows.ts [hunk 3]
```diff
@@ -47,237 +100,237 @@ const additionalFlow = (
...
 
-  We must merge the result of various calculations into a single Part.
-*/
-const mergeSourceFlows = (allParts: FlowPart[], sourceFlow: FlowPart[]): FlowPart[] =>
-  sourceFlow.reduce(
-    (acc, part) => {
-      const existing = allParts.find(p => isSamePart(p, part));
-      if (existing) {
-        existing.calculated = combineFlows(existing.calculated, part.calculated);
-        existing.liquid = combineLiquids(existing.liquid, part.liquid);
```
Comment (steersbob) [inline src/plugins/spark/features/ProcessView/calculateFlows.ts:189]: combineLiquids is unused

File: src/plugins/spark/features/ProcessView/parts/Pump.vue [hunk 4]
```diff
@@ -25,7 +25,7 @@ export default class Pump extends PartComponent {
         <line x1="7" y1="12.1" x2="-7" y2="-12.1"/>
         <!-- eslint-disable vue/attribute-hyphenation -->
         <animateTransform
-          v-if="!part.disabled"
+          v-if="!part.settings.disabled"
```
Comment (steersbob) [inline src/plugins/spark/features/ProcessView/parts/Pump.vue:39]: part.settings is undefined

File: src/plugins/spark/features/ProcessView/state.d.ts [hunk 5]
```diff
@@ -10,31 +10,27 @@ export interface Transitions {
...
 export interface PersistentPart {
   type: string;
   x: number;
   y: number;
   rotate: number;
-  liquidSource?: string;
-  closed?: boolean;
-  disabled?: boolean;
   flipped?: boolean;
+  settings?: Record<string, any>;
```
Comment (steersbob) [inline src/plugins/spark/features/ProcessView/state.d.ts]: - Why is this not meaningfullly typed?
- Why is it optional? This requires anyone accessing `part.settings` to build constructs like `part.settings || {}`.

**Review by steersbob — CHANGES_REQUESTED** [review]

**Review by steersbob — APPROVED** [review]

File: src/plugins/spark/features/ProcessView/state.d.ts (hunk 5 above)
Comment (elcojacobs) [reply src/plugins/spark/features/ProcessView/state.d.ts]: This will be the generic place for custom settings for each part. It is generic on purpose

**Comment by elcojacobs** [comment]
I think almost all issues are fixed now and this can be merged

**Pull request merged**